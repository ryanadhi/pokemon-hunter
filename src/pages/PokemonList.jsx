/* eslint-disable jsx-a11y/click-events-have-key-events */
/* eslint-disable jsx-a11y/no-noninteractive-element-interactions */
/* eslint-disable no-unused-vars */
import React, { useEffect, useState } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { useHistory } from 'react-router-dom';
import { getPokemons } from '../stores/actions/pokemon';

export default function PokemonList() {
  const [offset, setOffset] = useState(0);
  const [limit, setLimit] = useState(20);
  const dispatch = useDispatch();
  const history = useHistory();
  const { pokemons, myPokemon } = useSelector(
    (state) => state.pokemon,
  );

  useEffect(() => {
    dispatch(getPokemons(offset, limit));
  }, [offset, limit]);

  const countPokemonOwned = (name) => {
    const samePokemonOwned = myPokemon.data.filter((el) => el.name === name);
    if (samePokemonOwned.length === 0) {
      return '';
    }
    return `(${samePokemonOwned.length})`;
  };

  const nextButton = () => {
    setOffset(offset + limit);
  };
  const prevButton = () => {
    setOffset(offset - limit);
  };
  const goToDetail = (name) => {
    history.push(`/pokemon/${name}`);
  };

  return (
    <>
      <div>
        {pokemons.status === 'loading' && <p>loading...</p>}
        {pokemons.status === 'loaded' && pokemons.data.results.map((pokemon) => (
          <p onClick={() => goToDetail(pokemon.name)}>
            {pokemon.name}
            {countPokemonOwned(pokemon.name)}
          </p>
        ))}
      </div>
      {pokemons.data.previous && <button onClick={prevButton} type="button">prev</button>}
      {pokemons.data.next && <button onClick={nextButton} type="button">next</button>}
    </>
  );
}
