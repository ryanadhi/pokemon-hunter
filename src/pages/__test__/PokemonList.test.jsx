/* eslint-disable no-undef */
import React from 'react';
import { shallow } from 'enzyme';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';
import { render } from '@testing-library/react';
import { BrowserRouter } from 'react-router-dom';
import thunk from 'redux-thunk';
import PokemonList from '../PokemonList';

describe('PokemonList', () => {
  const initialState = {
    pokemon: {
      pokemons: {
        status: 'loaded',
        error: null,
        data: {
          count: 150,
          next: 'https://pokeapi.co/api/v2/pokemon?offset=20&limit=20',
          previous: null,
          results: [
            {
              id: 1,
              name: 'bulbasaur',
              url: 'https://pokeapi.co/api/v2/pokemon/1/',
            },
          ],
        },
      },
      pokemon: {
        status: 'loaded',
        error: null,
        data: {},
      },
      myPokemon: {
        data: [{
          id: 1,
          pokemon_id: 1,
          name: 'bulbasaur',
          nickname: 'bulbasaur-1',
        }],
      },
    },
  };
  const middlewares = [thunk];
  const mockStore = configureStore(middlewares);
  const store = mockStore(initialState);
  it('should render correctly in "debug" mode', () => {
    const wrapper = shallow(
      <BrowserRouter>
        <Provider store={store}>
          <PokemonList debug />
        </Provider>
      </BrowserRouter>,
    );

    expect(wrapper).toMatchSnapshot();
  });
  it('should render data from my pokemon', () => {
    const wrapper = render(
      <BrowserRouter>
        <Provider store={store}>
          <PokemonList debug />
        </Provider>
      </BrowserRouter>,
    );

    expect(wrapper.getByText('bulbasaur')).toBeInTheDocument();
  });
});
