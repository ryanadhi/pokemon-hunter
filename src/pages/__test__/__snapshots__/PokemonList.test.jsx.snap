// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PokemonList should render correctly in "debug" mode 1`] = `
<Router
  history={
    Object {
      "action": "POP",
      "block": [Function],
      "createHref": [Function],
      "go": [Function],
      "goBack": [Function],
      "goForward": [Function],
      "length": 1,
      "listen": [Function],
      "location": Object {
        "hash": "",
        "pathname": "/",
        "search": "",
        "state": undefined,
      },
      "push": [Function],
      "replace": [Function],
    }
  }
>
  <Provider
    store={
      Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      }
    }
  >
    <PokemonList
      debug={true}
    />
  </Provider>
</Router>
`;
